src/commands/cmd_bulk_insert.c

```diff
@@ -57,43 +46,61 @@ void _MGraph_BulkInsert(void *args) {
     RedisModule_ThreadSafeContextLock(ctx);
 
     RedisModuleString **argv = context->argv + 1; // skip "GRAPH.BULK"
-    int argc = context->argc - 2; // skip "GRAPH.BULK [GRAPHNAME]"
     RedisModuleString *rs_graph_name = *argv++;
+    int argc = context->argc - 2; // skip "GRAPH.BULK [GRAPHNAME]"
     RedisModuleKey *key;
-    size_t nodes = 0;   // Number of nodes created.
-    size_t edges = 0;   // Number of edge created.
 
     char reply[1024] = {0}; // Prepare the Redis string response
     int len;
 
     GraphContext *gc = NULL;
 
-    // Type chosen to match Redis conversion function
-    long long final_node_count;
-    long long final_edge_count;
+    // Number of entities already created
+    size_t initial_node_count = 0;
+
+    // Number of entities being created in this query
+    // (declared as long longs to match Redis conversion function)
+    long long nodes_in_query;
+    long long relations_in_query;
+
     // Check the next to see if this is a starting query
+    bool initial_query = false;
     if (!strcmp(RedisModule_StringPtrLen(*argv, 0), "BEGIN")) {
         argv ++;
-        if (_BeginBulkInsert(ctx, rs_graph_name, &final_node_count, &final_edge_count, argv) != BULK_OK) {
+        argc --;
+        initial_query = true;
+        // Verify that graph does not already exist.
+        key = RedisModule_OpenKey(ctx, rs_graph_name, REDISMODULE_READ);
+        if (key) {
+            RedisModule_ReplyWithError(ctx, "Key was already in Redis database at beginning of bulk insert operation.");
             goto cleanup;
         }
-        argv += 3; // skip "BEGIN [NODE_COUNT] [EDGE_COUNT]"
-        argc -= 3; // skip "BEGIN [NODE_COUNT] [EDGE_COUNT]"
+    }
 
+    // Read the user-provided counts for nodes and edges in the current query.
+    if (RedisModule_StringToLongLong(*argv++, &nodes_in_query) != REDISMODULE_OK) {
+        RedisModule_ReplyWithError(ctx, "Error parsing node count.");
+        goto cleanup;
+    }
+
+    if (RedisModule_StringToLongLong(*argv++, &relations_in_query) != REDISMODULE_OK) {
+        RedisModule_ReplyWithError(ctx, "Error parsing relation count.");
+        goto cleanup;
+    }
+    argc -= 2; // already read node count and edge count
+
+    if (initial_query) {
         // Create graph and initialize its data stores.
-        gc = GraphContext_New(ctx, rs_graph_name, final_node_count, final_edge_count);
-        // Exit if graph creation failed
-        if (gc == NULL) {
-            RedisModule_ReplyWithError(ctx, "Failed to allocate space for graph.");
-            goto cleanup;
-        }
+        gc = GraphContext_New(ctx, rs_graph_name, nodes_in_query, relations_in_query);
+        assert(gc);
```

Comment:
Handle in a similar way to line 99

Reply:
We gracefully error on duplicate keys in a "BEGIN" query where `initial_query` is set above. This assert is a 'the impossible happened' kind of scenario.